### ISSUE TITLE
[Feat]: Unpublish stream

### ISSUE THREAD

**Issue by talhatsk** [apivideo/api.video-reactnative-live-stream#79]
### Version

2.0.0

### Environment that reproduces the issue

Device iPhone xs max iOS 16.1

### Use case description

Currently, the stop() method disconnects the stream but doesn't unpublish it from the server. How can we unpublish the stream?

### Proposed solution

There should be a method to unpublish the stream from server as well.

**Comment by ThibaultBee** [comment]
Hi,

I guess you are talking about RTMP Command message. I don't think there is an official `unpublish` message. Which command message do you have in mind? (FCUnpublish?)

**Comment by talhatsk** [comment]
> Hi,
> 
> I guess you are talking about RTMP Command message. I don't think there is an official `unpublish` message. Which command message do you have in mind? (`FCUnpublish`, `closeStream`?)
> 
> I am just curious, what server do you use? Because if the connection is closed. There is no need to explicit `unpublish`.
> 
> Do you have the same issue with Android?

and yes it is same for android

**Comment by ThibaultBee** [comment]
> > Hi,
> > I guess you are talking about RTMP Command message. I don't think there is an official `unpublish` message. Which command message do you have in mind? (`FCUnpublish`, `closeStream`?)
> > I am just curious, what server do you use? Because if the connection is closed. There is no need to explicit `unpublish`.
> > Do you have the same issue with Android?
> 
> its `FCUnpublish` command. so what happen is if the stream gets unpublished then we get webhooks from rtmp server and if we close the stream (which is happening right now), webhooks won't be triggered and we have to wait for the timeout. Can we get any method which can send `FCUnpublish` command

If the connection between the device and your server is lost, you won't get a `FCUnpublish` and your webhooks won't be triggered. In that case, you will face the same issue and there is nothing we can do about it.

**Comment by talhatsk** [comment]
> > > Hi,
> > > I guess you are talking about RTMP Command message. I don't think there is an official `unpublish` message. Which command message do you have in mind? (`FCUnpublish`, `closeStream`?)
> > > I am just curious, what server do you use? Because if the connection is closed. There is no need to explicit `unpublish`.
> > > Do you have the same issue with Android?
> > 
> > 
> > its `FCUnpublish` command. so what happen is if the stream gets unpublished then we get webhooks from rtmp server and if we close the stream (which is happening right now), webhooks won't be triggered and we have to wait for the timeout. Can we get any method which can send `FCUnpublish` command
> 
> If the connection between the device and your server is lost, you won't get a `FCUnpublish` and your webhooks won't be triggered. In that case, you will face the same issue and there is nothing we can do about it.

Yes that is ok, network outage may occur or peer disconnections. But that is usually rare.  
But we need intentional control to unpublish our stream to complete the flow.
So a method would be good enough to unpublish stream along with existing closing stream

**Comment by ThibaultBee** [comment]
Hi,

For the implementation, I would rather like that `stopStream` calls `FCUnpublish` automatically instead of adding a specific method.

**Comment by talhatsk** [comment]
> Hi,
> 
> For the implementation, I would rather like that `stopStream` calls `FCUnpublish` automatically instead of adding a specific method.

Yes, that would also work

**Comment by ThibaultBee** [comment]
I made a fix in the Android library.
Could you have a look at HaishinKit.swift (the iOS library)?

**Comment by talhatsk** [comment]
> HaishinKit.swift

it checks if FCUnpublish to send a call to server to unpublish
`if let fcPublishName {
            connection.call("FCUnpublish", responder: nil, arguments: fcPublishName)
        }`

**Comment by ThibaultBee** [comment]
Just to confirm. The argument in the `FCUnpublish` is going to be the provided `StreamKey`.
Is it what you expect?

**Comment by talhatsk** [comment]
> Just to confirm. The argument in the `FCUnpublish` is going to be the provided `StreamKey`. Is it what you expect?

yes thats correct

**Comment by ThibaultBee** [comment]
Could you test the iOS version of the `main` branch?

**Comment by talhatsk** [comment]
> Could you test the iOS version of the `main` branch?

sure, i have to test the `example` folder on the `main` branch?

**Comment by ThibaultBee** [comment]
Yes Please :)
You can run:
```
yarn && yarn example ios
```

**Comment by talhatsk** [comment]
> Could you test the iOS version of the `main` branch?

so i've tested the main branch code, unfortunately its behaving the same way as before. I'm not getting any webhooks fired as it should be on the call of unpublish. The patchset is not working i guess.

**Comment by ThibaultBee** [comment]
hmm.
I can see `FCUnpublish` in Wireshark in the iOS example.
See:
<img width="1528" alt="Screenshot 2024-06-11 at 16 50 16" src="https://github.com/apivideo/api.video-reactnative-live-stream/assets/37510686/16eeb5af-f226-43d8-ab0d-0f28bfe54dd8">

**Comment by xoraingroup** [comment]
Hi there,

  I was testing the main branch as well. The FCUnpublish() followed by CloseStream() does not delete the stream properly from rtmp server. RTMP server reports connection closed by peer which is not the intended effect. In term of unpublishing the stream `FCUnpublish` command must be followed by `deleteStream()` command. 

Here is the packet captured from OBS while starting and stopping.

<img width="850" alt="image" src="https://github.com/apivideo/api.video-reactnative-live-stream/assets/17129052/7855e3c6-b4b5-484b-9876-07f4e805ccf6">

Can you provide an updated build with above modification.

**Comment by talhatsk** [comment]
@ThibaultBee I guess that must be the issue, `closeStream()` is not killing the stream properly. can we replace or add deleteStream()?

**Comment by ThibaultBee** [comment]
Could you add an issue for this in HaishinKit?

**Comment by talhatsk** [comment]
> `FCUnpublish()` followed by `closeStream()`

reported. [Link](https://github.com/shogo4405/HaishinKit.swift/issues/1472)

**Comment by ThibaultBee** [comment]
Could you test the android version on `main` branch?

**Comment by ThibaultBee** [comment]
Have you made any progress on the iOS part?

**Comment by talhatsk** [comment]
> Have you made any progress on the iOS part?

it is in progress yet.

**Comment by talhatsk** [comment]
> Have you made any progress on the iOS part?

@ThibaultBee  [PR](https://github.com/shogo4405/HaishinKit.swift/pull/1524) is up for iOS haishinkit

**Comment by talhatsk** [comment]
> Have you made any progress on the iOS part?

deleteStream [PR](https://github.com/shogo4405/HaishinKit.swift/pull/1524) is merged @ThibaultBee. You can integrate it in  the package now.

**Comment by ThibaultBee** [comment]
Waiting for next HK release

**Comment by ThibaultBee** [comment]
Have you test the fix made for Android?
